File: ch6/priorityqueue/C/priorityQueue.c
#include<limits.h>
#include "heap.c"

int heapMaximum(struct heap *h) {
        return h->A[0];
}

int heapExtractMaximum(struct heap *h) {
        //if(h->heapSize < 1) {
                //return INT_MIN;
        //}
        
        int max = h->A[0];
        h->A[0] = h->A[h->heapSize-1];
        h->heapSize = h->heapSize - 1;
        maxHeapify(h, 0);
        return max;
}

void heapIncreaseKey(struct heap *h, int i, int key) {
        if(key < h->A[i]) {
                return;
        }
        
        h->A[i] = key;
        while(i > 0 && h->A[parent(i)] < h->A[i]) {
                swap(h, i, parent(i));
                i = parent(i);
        }
}

void maxHeapInsert(struct heap *h, int key) {
        h->heapSize = h->heapSize + 1;
        h->A[h->heapSize-1] = INT_MIN;
        heapIncreaseKey(h, h->heapSize-1, key);
}

int main() {
        struct heap h;
        h.heapSize = 0;
        h.length = 0;

        maxHeapInsert(&h, 1);
        maxHeapInsert(&h, 2);
        maxHeapInsert(&h, 5);
        maxHeapInsert(&h, 4);

        printArray(&h);
        printf("%d\n", heapExtractMaximum(&h));
        printf("%d\n", heapExtractMaximum(&h));
}

File: ch6/priorityqueue/Java/PriorityQueue.java
public class PriorityQueue {

        private int[] A;
        private int heapSize;
        private static final int MAXSIZE = 100;

        public PriorityQueue() {
            A = new int[MAXSIZE];
            heapSize = 0;
        }

        public int heapMaximum() {
               return A[0];
        }

        public int heapExtractMax() {
                if(heapSize < 1) {
                        System.out.println("heap underflow");
                }

                int max = A[0];
                A[0] = A[heapSize-1];
                heapSize--;
                maxHeapify(0);
                return max;
        }

        public void heapIncreaseKey(int i, int key) {
                if(key < A[i]) {
                        System.out.println("new key is smaller than current key");
                }

                A[i] = key;
                while(i > 0 && A[parent(i)] < A[i]) {
                        swap(A, i, parent(i));
                        i = parent(i);
                }
        }

        public void maxHeapInsert(int key) {
                heapSize++;
                A[heapSize-1] = Integer.MIN_VALUE;
                heapIncreaseKey(heapSize-1, key);
        }
                      

        private void buildMaxHeap() {
                for(int i = A.length/2-1; i >= 0; i--) {
                        maxHeapify(i);
                }
        }

        private void maxHeapify(int i) {
                int largest = i;
                int l = left(i);
                int r = right(i);

                if(l < heapSize && A[l] > A[largest]) largest = l;
                
                if(r < heapSize && A[r] > A[largest]) largest = r;

                if(largest != i) {
                        swap(A, i, largest);
                        maxHeapify(largest);
                }
        }

        private void swap(int[] A, int i, int j) {
                int tmp = A[i];
                A[i] = A[j];
                A[j] = tmp;
        }

        private int parent(int i) {
                return i/2;
        }

        private int left(int i) {
                return 2*i+1;
        }

        private int right(int i) {
                return 2*i+2;
        }

        public String toString() {
                String out = "";
                for(int i = 0; i < heapSize; i++) {
                        out += A[i] + ", ";
                }
                out += "\n";
                return out;
        }

        public static void main(String[] args) {
                PriorityQueue pq = new PriorityQueue();

                pq.maxHeapInsert(1);
                pq.maxHeapInsert(5);
                pq.maxHeapInsert(4);
                pq.maxHeapInsert(6);

                System.out.print(pq.toString());

                System.out.println("Extracted: " + pq.heapExtractMax());
                System.out.println("Extracted: " + pq.heapExtractMax());
                System.out.println("Extracted: " + pq.heapExtractMax());
                System.out.println("Extracted: " + pq.heapExtractMax());
        }

}
